homeassistant/components/openrgb/__init__.py

```diff
@@ -0,0 +1,213 @@
+    }
+
+    # Initial device load
+    async def async_load_devices(device_list):
+        device_type_list = {}
+
+        for device in device_list:
+            ha_type = "light"
+            if ha_type not in device_type_list:
+                device_type_list[ha_type] = []
+            device_type_list[ha_type].append(device)
+
+            entity_id = orgb_entity_id(device)
+            if entity_id not in hass.data[DOMAIN]["entities"]:
+                hass.data[DOMAIN]["entities"][entity_id] = None
+
+        for ha_type, dev_ids in device_type_list.items():
+            config_entries_key = f"{ha_type}.openrgb"
+
+            if config_entries_key not in hass.data[DOMAIN][ENTRY_IS_SETUP]:
+                hass.data[DOMAIN]["pending"][ha_type] = dev_ids
+                hass.async_create_task(
+                    hass.config_entries.async_forward_entry_setup(entry, "light")
+                )
+                hass.data[DOMAIN][ENTRY_IS_SETUP].add(config_entries_key)
+            else:
+                async_dispatcher_send(
+                    hass, ORGB_DISCOVERY_NEW.format("light"), device_list
+                )
+
+    def _get_updated_devices():
+        try:
+            orgb.update()
+        except ConnectionError:
+            hass.data[DOMAIN]["connection_failed"]()
+            return None
+        return orgb.devices
+
+    await async_load_devices(_get_updated_devices())
+
+    async def async_poll_devices_update(event_time):
+        if not hass.data[DOMAIN]["online"]:
+            # try to reconnect
+            try:
+                hass.data[DOMAIN][ORGB_DATA].comms.start_connection()
+                hass.data[DOMAIN]["connection_recovered"]()
+            except ConnectionError:
+                hass.data[DOMAIN]["connection_failed"]()
+
+        device_list = await hass.async_add_executor_job(_get_updated_devices)
+
+        if device_list is None:
+            return
+
+        await async_load_devices(device_list)
+
+        newlist_ids = []
+        for device in device_list:
+            newlist_ids.append(orgb_entity_id(device))
+        for dev_id in list(hass.data[DOMAIN]["entities"]):
+            # Clean up stale devices, or alert them that new info is available.
+            if dev_id not in newlist_ids:
+                async_dispatcher_send(hass, SIGNAL_DELETE_ENTITY, dev_id)
+                hass.data[DOMAIN]["entities"].pop(dev_id)
+            else:
+                async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY, dev_id)
+
+    hass.data[DOMAIN][ORGB_TRACKER] = async_track_time_interval(
+        hass, async_poll_devices_update, TRACK_INTERVAL
+    )
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PULL_DEVICES, async_poll_devices_update
+    )
+
+    async def async_force_update(call):
+        """Force all devices to pull data."""
+        async_dispatcher_send(hass, SIGNAL_UPDATE_ENTITY)
+
+    hass.services.async_register(DOMAIN, SERVICE_FORCE_UPDATE, async_force_update)
```

Comment:
The difference between update and pull is quite unclear. Pull seems to also detect new devices. Could that always be done in update? Or could there be a parameter to update to do the extra work that pull does to improve clarity of the service(s) name?